File: chat-uikit/src/main/java/io/agora/chat/uikit/chat/interfaces/IChatLayout.java
package io.agora.chat.uikit.chat.interfaces;

import android.net.Uri;

import java.util.List;

import io.agora.chat.ChatMessage;
import io.agora.chat.MessageBody;
import io.agora.chat.uikit.chat.presenter.EaseHandleMessagePresenter;
import io.agora.chat.uikit.chat.widget.EaseChatInputMenu;
import io.agora.chat.uikit.chat.widget.EaseChatMessageListLayout;


public interface IChatLayout {

    /**
     * set external Presenter
     * @param presenter
     */
    void setPresenter(EaseHandleMessagePresenter presenter);
    /**
     * Get chat list
     * @return
     */
    EaseChatMessageListLayout getChatMessageListLayout();

    /**
     * Get input menu
     * @return
     */
    EaseChatInputMenu getChatInputMenu();

    /**
     * Get input content
     * @return
     */
    String getInputContent();

    /**
     * Whether to open the input monitoring
     * @param turnOn
     */
    void turnOnTypingMonitor(boolean turnOn);

    /**
     * Send text message
     * @param content
     */
    void sendTextMessage(String content);

    /**
     * Send text message
     * @param content
     * @param isNeedGroupAck Need group receipt
     */
    void sendTextMessage(String content, boolean isNeedGroupAck);

    /**
     * send @ message
     * @param content
     */
    void sendAtMessage(String content);

    /**
     * Send GIF message
     * @param name
     * @param identityCode
     */
    void sendBigExpressionMessage(String name, String identityCode);

    /**
     * Send voice message
     * @param filePath
     * @param length
     */
    void sendVoiceMessage(String filePath, int length);

    /**
     * Send voice message
     * @param filePath
     * @param length
     */
    void sendVoiceMessage(Uri filePath, int length);

    /**
     * Send picture message
     * @param imageUri
     */
    void sendImageMessage(Uri imageUri);

    /**
     * Send picture message
     * @param imageUri
     * @param sendOriginalImage
     */
    void sendImageMessage(Uri imageUri, boolean sendOriginalImage);

    /**
     * Send location message
     * @param latitude
     * @param longitude
     * @param locationAddress
     */
    void sendLocationMessage(double latitude, double longitude, String locationAddress);

    /**
     * Send video message
     * @param videoUri
     * @param videoLength
     */
    void sendVideoMessage(Uri videoUri, int videoLength);

    /**
     * Send a combine message.
     * @param message
     */
    void sendCombineMessage(ChatMessage message);

    /**
     * Send file message
     * @param fileUri
     */
    void sendFileMessage(Uri fileUri);

    /**
     * Add extension fields to the message
     * @param message
     */
    void addMessageAttributes(ChatMessage message);

    /**
     * Send message
     * @param message
     */
    void sendMessage(ChatMessage message);

    /**
     * Resend message
     * @param message
     */
    void resendMessage(ChatMessage message);

    /**
     * delete local message
     * @param message
     */
    void deleteMessage(ChatMessage message);

    /**
     * Delete local message list.
     * @param messages
     */
    void deleteMessages(List<String> messages);

    /**
     * Recall server message
     * @param message
     */
    void recallMessage(ChatMessage message);

    /**
     * modify message
     * @param messageId
     * @param messageBodyModified
     */
    void modifyMessage(String messageId, MessageBody messageBodyModified);

    /**
     * Set the edit message listening
     * @param listener
     */
    void setOnEditMessageListener(OnModifyMessageListener listener);

    void setOnChatLayoutListener(OnChatLayoutListener listener);

    /**
     * Used to monitor touch events for sending voice
     * @param voiceTouchListener
     */
    void setOnChatRecordTouchListener(OnChatRecordTouchListener voiceTouchListener);

    /**
     * Message withdrawal monitoring
     * @param listener
     */
    void setOnRecallMessageResultListener(OnRecallMessageResultListener listener);

    /**
     * Set the attribute event before sending message
     * @param sendMsgEvent
     */
    void setOnAddMsgAttrsBeforeSendEvent(OnAddMsgAttrsBeforeSendEvent sendMsgEvent);

    /**
     * The listener of reaction
     *
     * @param reactionListener
     */
    void setOnReactionListener(OnReactionMessageListener reactionListener);

    /**
     * Set the listener of multi-select.
     * @param listener  The listener of multi-select.
     */
    void setOnSelectClickListener(OnMessageSelectResultListener listener);
}

File: ease-im-kit/src/main/kotlin/com/hyphenate/easeui/feature/chat/interfaces/IChatLayout.kt
package com.hyphenate.easeui.feature.chat.interfaces

import android.net.Uri
import com.hyphenate.easeui.common.ChatMessage
import com.hyphenate.easeui.common.ChatMessageBody
import com.hyphenate.easeui.feature.chat.widgets.EaseChatInputMenu
import com.hyphenate.easeui.feature.chat.widgets.EaseChatMessageListLayout
import com.hyphenate.easeui.feature.thread.interfaces.OnMessageChatThreadClickListener
import com.hyphenate.easeui.feature.chat.widgets.EaseChatNotificationView
import com.hyphenate.easeui.viewmodel.messages.IChatViewRequest

interface IChatLayout {
    /**
     * Set the parent id for chat thread conversation.
     * @param parentId It is usually a group id which the chat thread belongs to.
     */
    fun setParentId(parentId: String?)

    /**
     * set custom ViewModel
     * @param viewModel
     */
    fun setViewModel(viewModel: IChatViewRequest?)

    /**
     * Get chat list
     * @return
     */
    val chatMessageListLayout: EaseChatMessageListLayout?

    /**
     * Get input menu
     * @return
     */
    val chatInputMenu: EaseChatInputMenu?

    /**
     * Get notification view
     */
    val chatNotificationView: EaseChatNotificationView?

    /**
     * Get input content
     * @return
     */
    val inputContent: String?

    /**
     * Whether to open the input monitoring
     * @param turnOn
     */
    fun turnOnTypingMonitor(turnOn: Boolean)

    /**
     * Dismiss the notification view.
     */
    fun dismissNotificationView(dismiss: Boolean)

    /**
     * Send text message
     * @param content
     * @param isNeedGroupAck Need group receipt
     */
    fun sendTextMessage(content: String?, isNeedGroupAck: Boolean = false)

    /**
     * send @ message
     * @param content
     */
    fun sendAtMessage(content: String?)

    /**
     * Send GIF message
     * @param name
     * @param identityCode
     */
    fun sendBigExpressionMessage(name: String?, identityCode: String?)

    /**
     * Send voice message
     * @param filePath
     * @param length
     */
    fun sendVoiceMessage(filePath: String?, length: Int)

    /**
     * Send voice message
     * @param filePath
     * @param length
     */
    fun sendVoiceMessage(filePath: Uri?, length: Int)

    /**
     * Send picture message
     * @param imageUri
     * @param sendOriginalImage
     */
    fun sendImageMessage(imageUri: Uri?, sendOriginalImage: Boolean = false)

    /**
     * Send location message
     * @param latitude
     * @param longitude
     * @param locationAddress
     */
    fun sendLocationMessage(latitude: Double, longitude: Double, locationAddress: String?)

    /**
     * Send video message
     * @param videoUri
     * @param videoLength
     */
    fun sendVideoMessage(videoUri: Uri?, videoLength: Int)

    /**
     * Send a combine message.
     * @param message
     */
    fun sendCombineMessage(message: ChatMessage?)

    /**
     * Send file message
     * @param fileUri
     */
    fun sendFileMessage(fileUri: Uri?)

    /**
     * Add extension fields to the message
     * @param message
     */
    fun addMessageAttributes(message: ChatMessage?)

    /**
     * Send message
     * @param message
     */
    fun sendMessage(message: ChatMessage?)

    /**
     * Resend message
     * @param message
     */
    fun resendMessage(message: ChatMessage?)

    /**
     * Report message
     * @param tag
     * @param reason
     * @param message
     */
    fun reportMessage(tag:String,reason:String,message: ChatMessage?)

    /**
     * delete local message
     * @param message
     */
    fun deleteMessage(message: ChatMessage?)

    /**
     * Delete local message list.
     * @param messages
     */
    fun deleteMessages(messages: List<String>?)

    /**
     * Recall server message
     * @param message
     */
    fun recallMessage(message: ChatMessage?)

    /**
     * modify message
     * @param messageId
     * @param messageBodyModified
     */
    fun modifyMessage(messageId: String?, messageBodyModified: ChatMessageBody?)

    /**
     * Set the edit message listening
     * @param listener
     */
    fun setOnEditMessageListener(listener: OnModifyMessageListener?)

    /**
     * Set the report message listening
     * @param listener
     */
    fun setOnReportMessageListener(listener:OnReportMessageListener?)

    /**
     * Set the translation message listening
     * @param listener
     */
    fun setOnTranslationMessageListener(listener: OnTranslationMessageListener?)

    /**
     * Set the thread view click listening
     * @param listener
     */
    fun setOnMessageThreadViewClickListener(listener: OnMessageChatThreadClickListener?)

    fun setOnChatLayoutListener(listener: OnChatLayoutListener?)

    /**
     * Used to monitor touch events for sending voice
     * @param voiceTouchListener
     */
    fun setOnChatRecordTouchListener(voiceTouchListener: OnChatRecordTouchListener?)

    /**
     * Message withdrawal monitoring
     * @param listener
     */
    fun setOnRecallMessageResultListener(listener: OnRecallMessageResultListener?)

    /**
     * Set the attribute event before sending message
     * @param onWillSendMessageListener
     */
    fun setOnWillSendMessageListener(onWillSendMessageListener: OnWillSendMessageListener?)

    /**
     * The listener of reaction
     *
     * @param reactionListener
     */
    fun setOnReactionListener(reactionListener: OnReactionMessageListener?)

    /**
     * Set layer finish listener.
     */
    fun setOnChatFinishListener(listener: OnChatFinishListener?)

    /**
     * Set chat presence listener.
     */
    fun setChatPresenceListener(listener: OnChatPresenceListener?)

}